unified-xen: backport unified EFI loader from xen-staging

osresearch opened the pull request:
This patch adds support for bundling the xen.efi hypervisor, the xen.cfg
configuration file, the Linux kernel and initrd, as well as the XSM,
and architectural specific files into a single "unified" EFI executable.
This allows an administrator to update the components independently
without requiring rebuilding xen, as well as to replace the components
in an existing image.

The resulting EFI executable can be invoked directly from the UEFI Boot
Manager, removing the need to use a separate loader like grub as well
as removing dependencies on local filesystem access.  And since it is
a single file, it can be signed and validated by UEFI Secure Boot without
requring the shim protocol.

It is inspired by systemd-boot's unified kernel technique and borrows the
function to locate PE sections from systemd's LGPL'ed code.  During EFI
boot, Xen looks at its own loaded image to locate the PE sections for
the Xen configuration (`.config`), dom0 kernel (`.kernel`), dom0 initrd
(`.ramdisk`), and XSM config (`.xsm`), which are included after building
xen.efi using objcopy to add named sections for each input file.

For x86, the CPU ucode can be included in a section named `.ucode`,
which is loaded in the efi_arch_cfg_file_late() stage of the boot process.

On ARM systems the Device Tree can be included in a section named
`.dtb`, which is loaded during the efi_arch_cfg_file_early() stage of
the boot process.

Note that the system will fall back to loading files from disk if
the named sections do not exist. This allows distributions to continue
with the status quo if they want a signed kernel + config, while still
allowing a user provided initrd (which is how the shim protocol currently
works as well).

Signed-off-by: Trammell hudson <hudson@trmm.net>

osresearch commented:
Key updated on key server, maybe ready for review.

osresearch commented:
Is the build error spurious or did something break since the last commit?
```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```

![image](https://user-images.githubusercontent.com/3068843/97352039-d901d580-1892-11eb-9ae0-51d388f08b36.png)
The key error might be caused by almost every key server showing failed status on `khps`?

marmarek commented:
> Is the build error spurious or did something break since the last commit?

It's unrelated to this PR.

>  The key error might be caused by almost every key server showing failed status on `khps`?

No, it's because your key on the keyservers is expired.

marmarek commented:
Please add the patches to xen.spec.in, otherwise they aren't be applied.

marmarek commented on patch-unified-0007-efi-Enable-booting-unified-hypervisor-kernel-initrd-.patch:
```diff
@@ -0,0 +1,559 @@
++            continue;
++        }
++
++#if 0
++        if ( image_size < sect->VirtualSize + sect->VirtualAddress )
++            blexit(L"PE invalid section size + address");
++#endif
```
Why disable it here?

osresearch replied:
The `blexit()` function isn't available here in this 4.14.  It is `static` in `arch/x86/efi/boot.c` and was added to the header `efi.h` in https://gitlab.com/xen-project/xen/-/commit/8856a914bffd56c018a794c06d4abb6018aa958d

osresearch replied:
I've added an additional patch to make `blexit()` available in `efi.h` to the backport series.

marmarek commented:
I think you should include "EFI: further "need_to_free" adjustments" patch too.

osresearch commented:
> Please add the patches to xen.spec.in, otherwise they aren't be applied.

Oops.  I added them to the `backports` region of the patches.

> I think you should include "EFI: further "need_to_free" adjustments" patch too.

Added the additional patch.

marmarek commented:
According to CI, patch failed to apply:
```
DEBUG: + /usr/bin/cat /builddir/build/SOURCES/patch-unified-0002-EFI-some-easy-constification.patch
DEBUG: + /usr/bin/patch -p1 -s --fuzz=0 --no-backup-if-mismatch
DEBUG: 1 out of 3 hunks FAILED -- saving rejects to file xen/common/efi/boot.c.rej
```

osresearch commented:
Patched up the patch to apply after the other EFI patches in the `xen.spec.in`; it now differs from the one in the xen staging tree, although some of the others do, too.

I'm still not sure what is wrong with my signing key -- I made sure it was unexpired and [uploaded it as directed](https://www.qubes-os.org/doc/code-signing/#upload-the-key) 

```
% gpg2 --edit-key 0F948052DDECBE68
sec  rsa4096/B65BFE540DEF86C0
     created: 2010-09-12  expires: 2021-12-04  usage: SC  
     trust: ultimate      validity: ultimate
ssb  rsa4096/87BA9505F961C525
     created: 2010-09-12  expires: 2021-12-04  usage: E   
ssb  rsa2048/0F948052DDECBE68
     created: 2014-11-15  expires: 2021-12-04  usage: S   
ssb  rsa2048/DA48704571ACDB80
     created: 2014-11-15  expires: 2021-12-04  usage: E   
[ultimate] (1). Trammell Hudson <trammell.hudson@gmail.com>
[ultimate] (2)  Trammell Hudson <hudson@swcp.com>
[ultimate] (3)  Trammell Hudson <hudson@trmm.net>
[ultimate] (4)  Trammell Hudson <hudson@osresearch.net>
% gpg2 --send-keys --keyserver pool.sks-keyservers.net  B65BFE540DEF86C0
gpg: sending key B65BFE540DEF86C0 to hkp://pool.sks-keyservers.net
%
```

marmarek commented:
Signature checker couldn't contact keyserver, as apparently some in the sks-keyserver.net pool have expired certificates... Should be better now.

fepitre commented:
PipelineRetry

fepitre commented:
PipelineRetry

osresearch commented:
The `build:vm-archlinux` job failed with "_This job depends on other jobs with expired/erased artifacts_". Is that something that I broke in this PR or an outside issue?

marmarek commented:
No, that's unrelated. dom0 build succeeded.

osresearch commented:
What further testing needs to be done to validate that the unified kernel patches don't break anythiing? Is it possible to download the artifacts for running in qemu or on real hardware?

marmarek commented:
Basic tests seems to be fine, but the above test suite doesn't include EFI boot. I'll launch another one with EFI boot tool.

marmarek commented:
Test suite including UEFI boots is also fine: https://openqa.qubes-os.org/tests/overview?distri=qubesos&version=4.1&build=20201218-4.1&groupid=1

osresearch commented:
Is there a way to download the artifacts that it built? I don't have a fully working qubes builder, but would like to test the resulting image to be sure that the unification process works.

marmarek commented:
There are two options:
- download packages from https://gitlab.com/QubesOS/qubes-vmm-xen/-/jobs/891462871/artifacts/browse/qubes-packages-mirror-repo/dom0-fc32/rpm/
- download full disk image of installed system with them (to be used within KVM with nested virt): https://openqa.qubes-os.org/tests/14517#downloads

osresearch commented:
Using the `boot/efi/EFI/qubes/xen-4.14.0.efi` from the package as `BOOTX64.EFI`, I get a crash when booting under qemu:
```
Xen 4.14.0 (c/s ) EFI loader
Using configuration file 'BOOTX64.cfg'
vmlinuz: 0x000000007c686000-0x000000007ccd5ca0
initrd.img: 0x000000007ab69000-0x000000007bf35b1c
!!!! X64 Exception Type - 0D(#GP - General Protection)  CPU Apic ID - 00000000 !!!!
ExceptionData - 0000000000000000
RIP  - 000000007D021205, CS  - 0000000000000038, RFLAGS - 0000000000010016
RAX  - 2494894D4FFC4D17, RCX - 000000009F850FD2, RDX - 000000007FF27FFF
RBX  - 000000007F9EE018, RSP - 000000007FF24630, RBP - 000000007D730360
RSI  - 000000007D448DB8, RDI - 000000007D618698
R8   - 0000000000000001, R9  - 000000007D62E100, R10 - 00000000000018C0
R11  - 0000000000000030, R12 - 000000007D21CAA3, R13 - 0000000000000212
R14  - 000000007D448DB8, R15 - 000000007FF24678
DS   - 0000000000000030, ES  - 0000000000000030, FS  - 0000000000000030
GS   - 0000000000000030, SS  - 0000000000000030
CR0  - 0000000080010033, CR2 - 0000000000000000, CR3 - 000000007FC01000
CR4  - 0000000000000668, CR8 - 0000000000000000
DR0  - 0000000000000000, DR1 - 0000000000000000, DR2 - 0000000000000000
DR3  - 0000000000000000, DR6 - 00000000FFFF0FF0, DR7 - 0000000000000400
GDTR - 000000007F9EEA98 0000000000000047, LDTR - 0000000000000000
IDTR - 000000007F4D4018 0000000000000FFF,   TR - 0000000000000000
FXSAVE_STATE - 000000007FF24290
!!!! Find image based on IP(0x7D021205)  (ImageBase=000000007CE00000, EntryPoint=000000007D43F0D0) !!!!
```

Booting that same config with Xen 4.15 (with the unified patches applied) works:
```
Xen 4.15-unstable (c/s Mon Aug 10 15:45:46 2020 +0100 git:7a4dd36-dirty) EFI loader
Using configuration file 'BOOTX64.cfg'
vmlinuz: 0x000000007c5d4000-0x000000007cc23ca0
initrd.img: 0x000000007ab69000-0x000000007bf35b1c
 __  __            _  _    _ ____                     _        _     _
 \ \/ /___ _ __   | || |  / | ___|    _   _ _ __  ___| |_ __ _| |__ | | ___
  \  // _ \ '_ \  | || |_ | |___ \ __| | | | '_ \/ __| __/ _` | '_ \| |/ _ \
  /  \  __/ | | | |__   _|| |___) |__| |_| | | | \__ \ || (_| | |_) | |  __/
 /_/\_\___|_| |_|    |_|(_)_|____/    \__,_|_| |_|___/\__\__,_|_.__/|_|\___|

(XEN) Xen version 4.15-unstable (hudson@) (gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0) debug=y  Wed Aug 12 19:12:34 CEST 2020
(XEN) Latest ChangeSet: Mon Aug 10 15:45:46 2020 +0100 git:7a4dd36-dirty
(XEN) build-id: fc561dd68dcfc8ce9ad90cbf161ceb16
(XEN) Bootloader: EFI
(XEN) Command line: console=com1 efi=attr=uc no-real-mode
....
```

This is without attempting to unify the kernel; I'm still attempting to track down where the fault is.

marmarek commented:
Interesting, this makes we wonder whether I've used correct package for the other tests...

marmarek commented:
Possibly important detail: Qubes 4.1 with EFI uses grub2-efi and boots through multiboot2+efi. This is what the test used.

osresearch commented:
It looks like this is a problem with the way I'm booting / testing.  With a clean version of the Qubes xen-4.14.0 patches (d24338d53c811973d4bff3c48ca5a282c05eaee6, which is one without my unified patches applied), the GPF still occurs.   These tests are not using grub2-efi -- they are booting the hypervisor directly from the UEFI (OVMF) firmware in qemu.   Diffing the configs and adding some tracing, it appears that the fault happens in `efi_exit_boot()` due to `CONFIG_EFI_SET_VIRTUAL_ADDRESS_MAP`.  Turning that off allows the vmm to make progress, although it fails entering the kernel with:

```
(XEN) Freed 2048kB init memory
mapping kernel into physical memory
about to get started...
Xen hypervisor allocated kernel memory conflicts with E820 map
(XEN) Hardware Dom0 halted: halting machine
```

A clean checkout of xen-4.14.0 with no patches applied, and built using the the Qubes `xen-4.14.0.config` still boots all the way into the dom0 kernel (including setting the EFI virtual address map), so something in a different patch seems to be conflicting with this boot technique.  I haven't tried digging into the patches to see why the clean 4.14.0 works, but the (non-unified) patched one fails yet. Should I split that into a separate issue?

marmarek commented:
> Should I split that into a separate issue?

Yes, makes sense. Can you include also full qemu cmdline?

osresearch commented:
Thanks for the help in diagnosing the OVMF errors in https://github.com/QubesOS/qubes-issues/issues/6298 -- turning off *both* `mapbs=1` and `exitbootservices=1` in the `xen.cfg` allowed the unified image to boot all the way into the dom0.

The command I used to generate the unified image was;
```
objcopy \
	--add-section .config=xen.cfg \
	--change-section-vma .config=0xffff82d041000000 \
	--add-section .kernel=vmlinuz \
	--change-section-vma .kernel=0xffff82d041100000 \
	--add-section .ramdisk=initrd.img \
	--change-section-vma .ramdisk=0xffff82d042000000 \
	xen.efi \
	boot/EFI/BOOT/BOOTX64.efi
```
Note that you need an up-to-date binutils to have a non-buggy `objcopy` since [the `BUILD_ID` alignment patch](https://lists.xenproject.org/archives/html/xen-devel/2020-09/msg00429.html) was not merged.

marmarek commented:
PipelineRetry
